[inductor] Add an AOT compilation mode for Inductor CPP backend

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94822

Summary: The AOT mode currently works for the CPP backend. When turned on, Inductor compiles the model code into a .so file with aot_inductor_entry as the entry function. If the AOT compilation fails, Inductor will explicitly fail.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

desertfire committed:
[WIP][inductor] Add an AOT compilation mode for Inductor

Summary: The AOT mode currently works for the CPP backend. When turned
on, Inductor compiles the model code into a .so file with
__aot_inductor_entry as the entry function.
If the AOT compilation fails, Inductor will explicitly fail.

[ghstack-poisoned]

desertfire committed:
Update on "[WIP][inductor] Add an AOT compilation mode for Inductor"


Summary: The AOT mode currently works for the CPP backend. When turned
on, Inductor compiles the model code into a .so file with
__aot_inductor_entry as the entry function.
If the AOT compilation fails, Inductor will explicitly fail.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
Output from running `test.sh`,

```
Turning on aten_graph for aot_inductor
[2023-02-14 15:12:29,045] torch._inductor.compile_fx: [INFO] Step 3: torchinductor compiling FORWARDS graph 0

#include "/tmp/torchinductor_binbao/cd/ccdu23xgmx4kl3rilvo5rfytlffccwsazjbxkv4urqfixqspbwj4.h"
extern "C" void kernel_cpp_0(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0,
                       float* __restrict__ out_ptr1)
{
    {
        for(long i0=0; i0<512; i0+=1)
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + 16*i0);
            auto tmp1 = tmp0.sin();
            auto tmp2 = decltype(tmp1)(1)/(decltype(tmp1)(1) + tmp1.neg().exp());
            auto tmp3 = tmp0.cos();
            auto tmp4 = decltype(tmp3)(1)/(decltype(tmp3)(1) + tmp3.neg().exp());
            tmp2.store(out_ptr0 + 16*i0);
            tmp4.store(out_ptr1 + 16*i0);
        }
        #pragma omp simd simdlen(8) 
        for(long i0=8192; i0<8192; i0+=1)
        {
            auto tmp0 = in_ptr0[i0];
            auto tmp1 = std::sin(tmp0);
            auto tmp2 = std::exp(-tmp1);
            auto tmp3 = 1 / (1 + tmp2);
            auto tmp4 = std::cos(tmp0);
            auto tmp5 = std::exp(-tmp4);
            auto tmp6 = 1 / (1 + tmp5);
            out_ptr0[i0] = tmp3;
            out_ptr1[i0] = tmp6;
        }
    }
}
std::vector<at::Tensor> __aot_inductor_entry(std::vector<at::Tensor> args) {
    at::Tensor arg0_1;
    arg0_1 = args[0];
    auto buf0 = at::empty_strided({8, 4, 16, 16}, {1024, 256, 16, 1}, at::ScalarType::Float); 
    auto buf1 = at::empty_strided({8, 4, 16, 16}, {1024, 256, 16, 1}, at::ScalarType::Float); 
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    arg0_1.reset();
    return std::vector<at::Tensor>({buf0, buf1});
}

[2023-02-14 15:12:35,669] torch._inductor.codecache: [INFO] AOT-Inductor compiles code into: /scratch/binbao/work/pytorch/test/inductor/aot/build/aot_inductor_output.so
```

desertfire committed:
Update on "[WIP][inductor] Add an AOT compilation mode for Inductor"


Summary: The AOT mode currently works for the CPP backend. When turned
on, Inductor compiles the model code into a .so file with
__aot_inductor_entry as the entry function.
If the AOT compilation fails, Inductor will explicitly fail.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -512,13 +512,32 @@ def guard_export_print(guards):
+    if aot_inductor:
+        from torch._inductor.decomposition import decompositions
+
+        if not aten_graph:
+            log.warning("Turning on aten_graph for aot_inductor")
+            aten_graph = True
+        decomposition_table = (
+            {**decomposition_table, **decompositions}
+            if decomposition_table is not None
+            else decompositions
+        )
```
Is this needed? Won't inductor call make_fx internally?

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -512,13 +512,32 @@ def guard_export_print(guards):
+    aten_graph=False,
+    decomposition_table=None,
+    tracing_mode="real",
+    aot_inductor=False,
```
Does this need to be a flag?  IMO something like:
```
graph, guards = dynamo.export(...)
compiled = inductor.compile_fx(graph)
```
is easier to understand and hack on.   For example, you could run some custom passes before calling inductor.

voznesenskym replied:
+1 - this should be fully after export

jansel commented on torch/_inductor/codecache.py:
```diff
@@ -496,7 +497,25 @@ def load(cls, source_code):
-                output_path = input_path[:-3] + "so"
+                output_path = f"{input_path[:-3]}.so"
+
+                if config.aot_codegen:
+                    from .codegen.wrapper import CppWrapperCodeGen
+
+                    assert (
+                        cls.output_path is None
+                    ), "AOT compilation should only generate a single .so file"
+                    cls.output_path = output_path = os.path.join(
+                        os.getcwd(), f"{config.aot_codegen_output_prefix}.so"
+                    )
+                    output_header = os.path.join(
+                        os.getcwd(), f"{config.aot_codegen_output_prefix}.h"
+                    )
+                    with open(output_header, "w") as header_file:
+                        header_file.writelines("#include <torch/torch.h>\n\n")
+                        header_file.writelines(f"{CppWrapperCodeGen.decl_str};\n")
+                    log.info(f"AOT-Inductor compiles code into: {output_path}")
+
```
I'd suggest moving this logic somewhere else.  Changing the behavior of `load()` to have side effects like this is confusing.

jansel commented on torch/_inductor/config.py:
```diff
@@ -13,6 +13,20 @@
 # Whether to enable printing the source code for each future
 verbose_progress = False
 
+# AOT compile model into .so file
+aot_codegen = False
+
+# Name for generated .h and .so files
+aot_codegen_output_prefix = "aot_inductor_output"
+
+
+def set_aot_codegen():
+    global aot_codegen
+    aot_codegen = True
+    global debug
+    debug = True
+
+
```
I don't like the UX of using `config` to change behavior.

I'd suggest adding a new function like `aot_compile_fx(graph, example_inputs)` instead.  You can put mode flags on V.graph rather than in the config.

jansel reviewed (changes requested):
How are model weights handled by this?

voznesenskym commented:
Discussed offline a bit - but you don't need to make it part of export. You can make AOTInductor a purely additive thing. 

One thing we could do that would be minimal changes is to not change how we produce a module or wrapper today under cpp_wrapper but instead make this aot thing purely additive - produce a side artifact of the .so (as we do), and a .h (As we do not yet do), if a flag is set. This would allow us to button up an impl in very flew lines, at the cost of a little bit of redundant work.

Check out how we do

```
if self._can_use_cpp_wrapper:
    self.wrapper_code = CppWrapperCodeGen()
```
You can do something like
```
if self.aot
```
And provide your own AOTCodeGen alongside the other CodeGen (either composition, or add support to emit multiple codegen?). You can then use the flag to produce a pure compiled artifact as a side effect - no need to change any of the mainline flow of compilation. 

Once you have that - you can just call compile_fx after export, with the aot flag enabled.

desertfire committed:
Update on "[WIP][inductor] Add an AOT compilation mode for Inductor"


Summary: The AOT mode currently works for the CPP backend. When turned
on, Inductor compiles the model code into a .so file with
__aot_inductor_entry as the entry function.
If the AOT compilation fails, Inductor will explicitly fail.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

jansel reviewed (approved):
What is here looks good to me.

Still need to solve the issue of weight handling.

Should `__aot_inductor_entry` be just `aot_inductor_entry`, or perhaps a function name passed in by the user to avoid name conflicts?  If the user is intended to call it we shouldn't name it `__*`.

desertfire commented:
I will update with refactoring after https://github.com/pytorch/pytorch/pull/95594/ lands. Weight handling will come as the next PR.

desertfire committed:
Update on "[WIP][inductor] Add an AOT compilation mode for Inductor"


Summary: The AOT mode currently works for the CPP backend. When turned
on, Inductor compiles the model code into a .so file with
__aot_inductor_entry as the entry function.
If the AOT compilation fails, Inductor will explicitly fail.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Add an AOT compilation mode for Inductor CPP backend"


Summary: The AOT mode currently works for the CPP backend. When turned on, Inductor compiles the model code into a .so file with aot_inductor_entry as the entry function. If the AOT compilation fails, Inductor will explicitly fail.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge -r

desertfire committed:
Update on "[inductor] Add an AOT compilation mode for Inductor CPP backend"


Summary: The AOT mode currently works for the CPP backend. When turned on, Inductor compiles the model code into a .so file with aot_inductor_entry as the entry function. If the AOT compilation fails, Inductor will explicitly fail.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

jgong5 commented on test/inductor/aot/cpp/test.cpp:
```diff
@@ -0,0 +1,41 @@
+    for (const auto& pair : net.named_parameters()) {
+      inputs.push_back(pair.value());
+    }
+    torch::Tensor results_opt = aot_inductor_entry(inputs);
```
How do we plan to support the case where multiple AOT compiled modules are needed in a program? Would the same entry point cause symbol clashing?

desertfire replied:
Right now we are only dealing with single-graph scenario. We can update the naming later if needed.

desertfire commented:
@pytorchbot merge -f "CI failures are flaky"

clee2000 commented:
@pytorchbot revert -m "broke inductor_tmm_cpu_accuracy, https://hud.pytorch.org/pytorch/pytorch/commit/73b66098b2f43be508e1975fd6a425ed6308b993#11745396725" -c ignoredsignal

desertfire commented:
`xcit_large_24_p8_224` has shown a large memory consumption in the past which causes flaky OOM  failures (https://hud.pytorch.org/hud/pytorch/pytorch/master/1?per_page=50&name_filter=inductor_timm_cpu_acc). I think somehow this PR triggers the issue more consistently. I am not sure why though because the change here is not enabled by default.

I will take the test offline and create an issue for tracking.

desertfire commented:
@pytorchbot rebase

desertfire committed:
Update on "[inductor] Add an AOT compilation mode for Inductor CPP backend"


Summary: The AOT mode currently works for the CPP backend. When turned on, Inductor compiles the model code into a .so file with aot_inductor_entry as the entry function. If the AOT compilation fails, Inductor will explicitly fail.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

clee2000 commented:
Also, this PR seems to almost quadruple the time it takes to run inductor/test_torchinductor_opinfo (prev ~1hr, after ~3.75 hr)
